### PR TITLE
Adls data plane task add debug functionality

### PR DESCRIPTION
**Pull request by rahuldutta90**
## Description
Add debug functionality to ADLS dataplane cmdlets

## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#making-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr)
    - [x] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

### CODE REVIEW

File: src/ResourceManager/DataLakeStore/AzureRM.DataLakeStore.psd1 [hunk 1]
```diff
@@ -152,11 +152,7 @@ PrivateData = @{
 
         # ReleaseNotes of this module
         ReleaseNotes = '
-		* Export-AzureRmDataLakeStoreItem (https://github.com/Azure/azure-powershell/blob/adls-data-plane/src/ResourceManager/DataLakeStore/documentation/upcoming-breaking-changes.md) - Deprecated parameters PerFileThreadCount, ConcurrentFileCount and introduced parameter Concurrency
-		* Import-AzureRMDataLakeStoreItem (https://github.com/Azure/azure-powershell/blob/adls-data-plane/src/ResourceManager/DataLakeStore/documentation/upcoming-breaking-changes.md) -Deprecated parametersPerFileThreadCount, ConcurrentFileCount and introduced parameter Concurrency
-		* Get-AzureRMDataLakeStoreItemContent - Fixed the tail behavior for contents greater than 4MB
-		* Set-AzureRMDataLakeStoreItemExpiry - Introduced new parameter set SetRelativeExpiry for setting relative expiration time
-		* Remove-AzureRmDataLakeStoreItem (https://github.com/Azure/azure-powershell/blob/adls-data-plane/src/ResourceManager/DataLakeStore/documentation/upcoming-breaking-changes.md) - Deprecated parameter Clean.
+		* Add Debug functionality to Datalake store dataplane cmdlets
```
Comment (cormacpayne) [inline src/ResourceManager/DataLakeStore/AzureRM.DataLakeStore.psd1]: @rahuldutta90 you can revert the changes made to the `ReleaseNotes` -- we will update this metadata when we release this module again

File: src/ResourceManager/DataLakeStore/Commands.DataLakeStore/DataPlaneModels/DataLakeStoreFileSystemClient.cs [hunk 2]
```diff
@@ -45,8 +45,17 @@ public class DataLakeStoreFileSystemClient
...
         private const long NeverExpireValue = 253402300800000;
         internal const int ImportExportMaxThreads = 256;
+        private readonly LoggingConfiguration _adlsLoggerConfig;
+
         #region Constructors
 
+        static DataLakeStoreFileSystemClient()
+        {
+            // ConfigurationItemFactory.Default.Targets.RegisterDefinition("AdlsLogger", typeof(AdlsLoggerTarget));
+            Target.Register<Microsoft.Azure.Commands.DataLakeStore.Models.AdlsLoggerTarget>("AdlsLogger"); //generic
```
Comment (cormacpayne) [inline src/ResourceManager/DataLakeStore/Commands.DataLakeStore/DataPlaneModels/DataLakeStoreFileSystemClient.cs]: @rahuldutta90 what is the purpose of this target in `NLog`?

File: src/ResourceManager/DataLakeStore/Commands.DataLakeStore/DataPlaneModels/DataLakeStoreFileSystemClient.cs [hunk 3]
```diff
@@ -45,8 +45,17 @@ public class DataLakeStoreFileSystemClient
         private const int MaxConnectionLimit = 1000;
         private const long NeverExpireValue = 253402300800000;
         internal const int ImportExportMaxThreads = 256;
+        private readonly LoggingConfiguration _adlsLoggerConfig;
+
         #region Constructors
 
+        static DataLakeStoreFileSystemClient()
+        {
+            // ConfigurationItemFactory.Default.Targets.RegisterDefinition("AdlsLogger", typeof(AdlsLoggerTarget));
```
Comment (cormacpayne) [inline src/ResourceManager/DataLakeStore/Commands.DataLakeStore/DataPlaneModels/DataLakeStoreFileSystemClient.cs]: @rahuldutta90 nit: would you mind removing this line if it's going to be commented-out?

File: src/ResourceManager/DataLakeStore/Commands.DataLakeStore/DataPlaneModels/DataLakeStoreFileSystemClient.cs [hunk 4]
```diff
@@ -68,6 +77,22 @@ public DataLakeStoreFileSystemClient(IAzureContext context)
             ServicePointManager.UseNagleAlgorithm = false;
         }
 
+        public DataLakeStoreFileSystemClient(IAzureContext context, DataLakeStoreFileSystemCmdletBase cmdlet) : this(context)
+        {
+
```
Comment (cormacpayne) [inline src/ResourceManager/DataLakeStore/Commands.DataLakeStore/DataPlaneModels/DataLakeStoreFileSystemClient.cs]: @rahuldutta90 nit: please remove this blank line

File: src/ResourceManager/DataLakeStore/AzureRM.DataLakeStore.psd1 (hunk 1 above)
Comment (rahuldutta90) [reply src/ResourceManager/DataLakeStore/AzureRM.DataLakeStore.psd1]: @cormacpayne Reverted it.

File: src/ResourceManager/DataLakeStore/Commands.DataLakeStore/DataPlaneModels/DataLakeStoreFileSystemClient.cs (hunk 3 above)
Comment (rahuldutta90) [reply src/ResourceManager/DataLakeStore/Commands.DataLakeStore/DataPlaneModels/DataLakeStoreFileSystemClient.cs]: @cormacpayne  Removed. Thanks for catching it.

File: src/ResourceManager/DataLakeStore/Commands.DataLakeStore/DataPlaneModels/DataLakeStoreFileSystemClient.cs (hunk 4 above)
Comment (rahuldutta90) [reply src/ResourceManager/DataLakeStore/Commands.DataLakeStore/DataPlaneModels/DataLakeStoreFileSystemClient.cs]: @cormacpayne Removed it.

File: src/ResourceManager/DataLakeStore/Commands.DataLakeStore/DataPlaneModels/DataLakeStoreFileSystemClient.cs (hunk 2 above)
Comment (rahuldutta90) [reply src/ResourceManager/DataLakeStore/Commands.DataLakeStore/DataPlaneModels/DataLakeStoreFileSystemClient.cs]: @cormacpayne NLog is used by the ADLS dataplane sdk to log debug messages. We can create a custom target which basically queues the debug data to the ConcurrentQueue (the one available to this powershell framework) for debug messages. So whenever we use our nlog framework to write debug message all of them will get queued in the concurrent queue. I have put some more comments in the target class.

Comment (rahuldutta90) [reply src/ResourceManager/DataLakeStore/Commands.DataLakeStore/DataPlaneModels/DataLakeStoreFileSystemClient.cs]: @cormacpayne  https://github.com/NLog/NLog/wiki/Extending-NLog

File: src/ResourceManager/DataLakeStore/ChangeLog.md [hunk 5]
```diff
@@ -17,6 +17,14 @@
     * Overview of change #1
         - Additional information about change #1
 -->
+## Current Release
+
+* Add debug functionality
+
+## Version 5.1.1
```
Comment (markcowl) [inline src/ResourceManager/DataLakeStore/ChangeLog.md]: I don't think you want to change the order of this

File: src/ResourceManager/DataLakeStore/Commands.DataLakeStore/DataPlaneModels/DataLakeStoreFileSystemClient.cs [hunk 6]
```diff
@@ -495,28 +524,22 @@ public void AppendToFile(string filePath, string accountName, byte[] contents)
...
             var fileTarget = new FileTarget();
-            config.AddTarget("file", fileTarget);
+            _adlsLoggerConfig.AddTarget("file", fileTarget);
 
-            // Step 3. Set target properties 
             fileTarget.FileName = fileName;
-            fileTarget.Layout = "${message}";
+            
+            var rule = new LoggingRule("adls.dotnet.*", NLog.LogLevel.Debug, fileTarget);
+            _adlsLoggerConfig.LoggingRules.Add(rule);
```
Comment (markcowl) [inline src/ResourceManager/DataLakeStore/Commands.DataLakeStore/DataPlaneModels/DataLakeStoreFileSystemClient.cs]: Can you describe what this actually logs?  We want to make sure that request method and headers and response status and headers are included

Comment (rahuldutta90) [reply src/ResourceManager/DataLakeStore/Commands.DataLakeStore/DataPlaneModels/DataLakeStoreFileSystemClient.cs]: @markcowl This is the debug line NLog posts:

2018-02-16 09:23:59.1720|DEBUG|adls.dotnet.WebTransport|HTTPRequest,Succeeded,cReqId:cc6758eb-656f-4bd7-b589-67165664e899.0,lat:6496,err,Reqlen:0,Resplen:0,token_ns:2,sReqId:2551abc6-ed04-4a99-b71f-264da768cf7d,path:/NewFile,qp:op=CREATE&overwrite=True&leaseid=30038d32-172f-4054-ac08-359421c24282&filesessionid=30038d32-172f-4054-ac08-359421c24282&CreateParent=True&write=true&syncFlag=DATA&api-version=2017-08-01

Comment (rahuldutta90) [reply src/ResourceManager/DataLakeStore/Commands.DataLakeStore/DataPlaneModels/DataLakeStoreFileSystemClient.cs]: @markcowl For failure case only we have the httpstatus in response like here we have InternalServorError:
2018-02-14 20:02:48.5660|DEBUG|adls.dotnet.WebTransport|HTTPRequest,failed,cReqId:e1dc5c6d-a9c1-4202-8d99-7e6c00d19f8d.0,lat:723,errInternalServerError: RuntimeException,Reqlen:15,Resplen:0,token_ns:0,sReqId:a2ed5f89-f8a7-42e6-af1c-187806135668,path:/Test/dir1/testConcurrentAppendParallel_15,qp:op=CONCURRENTAPPEND&appendMode=autocreate&api-version=2017-08-01

File: src/ResourceManager/DataLakeStore/ChangeLog.md (hunk 5 above)
Comment (rahuldutta90) [reply src/ResourceManager/DataLakeStore/ChangeLog.md]: @markcowl Addressed it.

**Comment by maddieclayton** [comment]
@rahuldutta90 Please fix the merge conflicts that have arisen.

**Comment by cormacpayne** [comment]
@rahuldutta90 Hey Rahul, what's the status of this PR? I know that we were looking to re-format the debug log, as well as pull in the latest changes from the `preview` branch.

**Comment by rahuldutta90** [comment]
@cormacpayne I am in process of making the changes in SDK. Once it is released, I will integrate that version with this PR.

**Comment by markcowl** [comment]
on demand run here: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/636/
